weave/workers/weave_scorer.py

```diff
@@ -0,0 +1,489 @@
+    server = get_trace_server()
+    server.feedback_create(feedback_req)
+
+
+def _task_done_callback(
+    task: asyncio.Task,
+    messages: list[Message],
+    consumer: KafkaConsumer,
+    tasks: set[asyncio.Task],
+) -> None:
+    tasks.discard(task)
+
+    try:
+        task.result()
+        for msg in messages:
+            consumer.commit(msg)
+    except RetryError as e:
+        logger.warning("Retries exhausted. Messages should be re-consumed.")
+    except Exception as e:
+        logger.exception(
+            f"Error processing message: {e.__class__.__name__} {e}", exc_info=e
+        )
+
+
+def handle_kafka_errors(msg: Message) -> bool:
+    """Handle Kafka-specific errors."""
+    if msg.error():
+        if msg.error().code() == KafkaError._PARTITION_EOF:
+            logger.error(
+                "%% %s [%d] reached end at offset %d\n",
+                msg.topic(),
+                msg.partition(),
+                msg.offset(),
+            )
+        else:
+            logger.exception("Kafka error: %s", msg.error())
+
+        return False
+
+    return True
+
+
+async def cleanup_and_wait(tasks: set[asyncio.Task]) -> set[asyncio.Task]:
+    """Clean up completed tasks and wait for pending ones."""
+    done_tasks = {t for t in tasks if not t.done()}
+
+    if done_tasks:
+        _, pending = await asyncio.wait(done_tasks, return_when=asyncio.FIRST_COMPLETED)
+        return pending
+
+    return set()
+
+
+KAFKA_CONSUMER_GROUP_ID = "weave-worker-scorer"
+
+
+async def process_project_ended_calls(
+    project_id: str,
+    ended_calls: list[tsi.EndedCallSchemaForInsert],
+) -> None:
+    if len(ended_calls) == 0:
+        logger.warning("No ended calls to process, this should not happen")
+        return
+
+    project_id = ended_calls[0].project_id
+
+    active_monitors = get_active_monitors(project_id)
+
+    call_ids = [ended_call.id for ended_call in ended_calls]
+
+    for active_monitor in active_monitors:
+        monitor = active_monitor["monitor"]
+        monitor_internal_ref = active_monitor["internal_ref"]
+        wb_user_id = active_monitor["wb_user_id"]
+
+        # Here we potentially query the same calls multiple times.
+        # The reasons is that each monitor has its own call filter, so we need to query the DB for each monitor.
+        # We could merge the call filters and do a single query, but then
+        # we would need to figure out which call match what monitor filter.
+        # Currently we have no way to apply filters outside the DB
+        calls = await get_filtered_calls(
+            project_id,
+            call_ids,
+            monitor.op_names,
+            monitor.query,
+        )
+
+        with get_trace_server().call_batch():
+            for call in calls:
+                for scorer in monitor.scorers:
+                    logger.info("Applying scorer %s to call %s", scorer.name, call.id)
+                    await apply_scorer(
```

Comment:
highly recommend asyncio gather over these coroutines, otherwise you are going to serially process each call

Reply:
It actually does not matter at this time since `apply_scorer` is actually synchronous. It should actually not have an `async` on it. Right now it assumes all `Scorer.score` calls are sync. Of course when we introduce `LLMAsAJudgeScorer` the `score` call will be async. At that point there will be a benefit in making `apply_scorer` async.
I made the suggested change anyway in anticipation for a truly async `apply_scorer`.